subsys/bluetooth/common/bt_str.c

```diff
@@ -141,8 +144,18 @@ const char *bt_hci_err_to_str(uint8_t hci_err)
 	if (hci_err < ARRAY_SIZE(mapping_table) && mapping_table[hci_err]) {
 		return mapping_table[hci_err];
 	} else {
-		return "(unknown)";
+		static char err_as_hex[17];
+		sprintf(err_as_hex, "(unknown)(0x%02x)", hci_err);
+		return err_as_hex;
 	}
 
 	#undef HCI_ERR
 }
+#else
+const char *bt_hci_err_to_str(uint8_t err)
+{
+	static char err_as_hex[5];
```

Comment:
This has to at least be a thread local variable to be thread safe. We are committing to thread-safety as much as we can for new API.

Isn't this a bad pattern anyway? Returning a shared resource without a lock. Especially in public API, this can be very confusing for users. How do we document this API?

I think it's preferable to take the destination buffer as an argument, sprintf-like.

Reply:
To consider: We can make the function not allocate and instead take the destination buffer as input, a like `snprintf`. Unlike `snprintf`, we return the same pointer. This allows using the function directly in log statements. Finally, we add a macro that allocates an automatic variable for ease of use.

```c
#define BT_HCI_ERR_TO_STR(err) bt_hci_err_to_str(&(char[16]){}, 16, (err))
const char *bt_hci_err_to_str(char *dst, size_t dst_size, uint8_t err);
```

Reply:
That is a good solution, although it requires the allocation of that string before each log statement. Given that the resulting string is part of a limited set of possibilities, have a constant lookup table would still be better IMO, as we don't really need to do anything dynamically (unless we want "unknown(0xYY)").

Reply:
> I thought that it might be a good idea to add an error code when it is unknown to make it clear what is going wrong. Maybe we can simply log this separately?
> 
> ```c
> const char *bt_hci_err_to_str(uint8_t hci_err)
> {
>   	const char * const mapping_table[] = {ENTRIES};
> 	if (hci_err < ARRAY_SIZE(mapping_table) && mapping_table[hci_err]) {
> 		return mapping_table[hci_err];
> 	} else {
> 		LOG_ERR("Unknown HCI error code translation %d", hci_err);
> 		return "(unknown)";
> 	}
> }
> ```
> 
> Should we keep the `bt_hci_err_to_str()` variant that converts to a hex representation?

If we want to avoid that original suggestion with `%s (0x%02X)` then I think this would be the best solution, assuming that `mapping_table` here contains the entirety of what we want to get. 

Otherwise the original suggestion was never bad, just that it would be nice to be able to always just do `%s`